**Linux** [Build 8602](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=75997&buildTypeId=WFPR) outcome was **FAILURE** using a merge of 75758529bf26f3522707707355a3c2d38657fa25
Summary: Tests failed: 1 (1 new), passed: 3248, ignored: 363 Build time: 00:50:46
### Failed tests
```
org.jboss.as.test.integration.ejb.transaction.cmt.lifecyclenooptxs.LifecycleNoopTxTestCase.testStateless: java.lang.AssertionError: number of committed txs should be increased by 2 expected:<5> but was:<4>
	at org.jboss.as.test.integration.ejb.transaction.cmt.lifecyclenooptxs.LifecycleNoopTxTestCase.testStateless(LifecycleNoopTxTestCase.java:122)
------- Stdout: -------
 [0m01:08:08,607 INFO  [org.jboss.weld.deployer] (MSC service thread 1-3) WFLYWELD0010: Stopping weld service for deployment ejbLifecycleAnnotations.jar
 [0m [0m01:08:08,611 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-2) WFLYSRV0028: Stopped deployment ejbLifecycleAnnotations.jar (runtime-name: ejbLifecycleAnnotations.jar) in 6ms
 [0m [0m01:08:08,615 INFO  [org.jboss.as.repository] (management-handler-thread - 1) WFLYDR0002: Content removed from location /opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/basic/target/jbossas/standalone/data/content/36/bd0cbcf9f97599738582f3860c7121ce2cbb89/content
 [0m [0m01:08:08,615 INFO  [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0009: Undeployed "ejbLifecycleAnnotations.jar" (runtime-name: "ejbLifecycleAnnotations.jar")
 [0m [0m01:08:08,636 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0010: Unbound data source [java:jboss/datasources/batch]
 [0m [0m01:08:08,637 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0010: Unbound data source [java:jboss/datasources/ExampleDS]
 [0m [0m01:08:08,639 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-1) WFLYUT0019: Host default-host stopping
 [0m [0m01:08:08,644 INFO  [org.jboss.as.clustering.infinispan] (MSC service thread 1-2) WFLYCLINF0003: Stopped local-query cache from hibernate container
 [0m [0m01:08:08,648 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-3) WFLYSRV0028: Stopped deployment arquillian-service (runtime-name: arquillian-service) in 16ms
 [0m [0m01:08:08,663 INFO  [org.jboss.as.connector.deployers.jdbc] (MSC service thread 1-3) WFLYJCA0019: Stopped Driver service with driver-name = h2
 [0m [0m01:08:08,676 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-3) WFLYUT0008: Undertow HTTP listener default suspending
 [0m [0m01:08:08,682 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-3) WFLYUT0007: Undertow HTTP listener default stopped, was bound to [::1]:8080
 [0m [0m01:08:08,682 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-3) WFLYUT0004: Undertow 1.3.0.Final stopping
 [0m01:08:08,688 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-12) EJBCLIENT000016: Channel Channel ID f143b869 (outbound) of Remoting connection 0157e8e2 to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,688 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-12) EJBCLIENT000016: Channel Channel ID a084b280 (outbound) of Remoting connection 0157e8e2 to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,688 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-12) EJBCLIENT000016: Channel Channel ID c84c4d98 (outbound) of Remoting connection 0157e8e2 to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,689 INFO  [org.jboss.ejb.client.remoting] (Remoting "client" task-7) EJBCLIENT000016: Channel Channel ID 9ce87cbc (outbound) of Remoting connection 0001e48c to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,693 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-5) EJBCLIENT000016: Channel Channel ID 93ecd645 (outbound) of Remoting connection 0157e8e2 to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,694 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-6) EJBCLIENT000016: Channel Channel ID 97243224 (outbound) of Remoting connection 01da4b5f to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,694 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-6) EJBCLIENT000016: Channel Channel ID bd79e8ab (outbound) of Remoting connection 01da4b5f to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,694 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-6) EJBCLIENT000016: Channel Channel ID b346cb9c (outbound) of Remoting connection 01da4b5f to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,694 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-6) EJBCLIENT000016: Channel Channel ID a7f22bb9 (outbound) of Remoting connection 01da4b5f to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,696 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-15) EJBCLIENT000016: Channel Channel ID 8c6225c3 (outbound) of Remoting connection 01da4b5f to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,701 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-9) EJBCLIENT000016: Channel Channel ID 9c0a6844 (outbound) of Remoting connection 002db130 to /0:0:0:0:0:0:0:1:8080 can no longer process messages
01:08:08,702 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-5) EJBCLIENT000016: Channel Channel ID a95453d6 (outbound) of Remoting connection 01da4b5f to /0:0:0:0:0:0:0:1:8080 can no longer process messages
 [0m01:08:08,709 INFO  [org.jboss.as.mail.extension] (MSC service thread 1-3) WFLYMAIL0002: Unbound mail session [java:jboss/mail/Default]
 [0m [0m01:08:08,721 INFO  [org.jboss.as] (MSC service thread 1-2) WFLYSRV0050: WildFly Full 10.0.0.CR4-SNAPSHOT (WildFly Core 2.0.0.CR7) stopped in 88ms
 [0m [0m01:08:08,721 INFO  [org.jboss.as] (MSC service thread 1-2) WFLYSRV0049: WildFly Full 10.0.0.CR4-SNAPSHOT (WildFly Core 2.0.0.CR7) starting
 [0m [0m01:08:08,888 INFO  [org.jboss.as.server] (Controller Boot Thread) WFLYSRV0039: Creating http management service using socket-binding (management-http)
 [0m [0m01:08:08,892 INFO  [org.jboss.as.connector.subsystems.datasources] (ServerService Thread Pool -- 32) WFLYJCA0004: Deploying JDBC-compliant driver class org.h2.Driver (version 1.3)
 [0m [0m01:08:08,907 INFO  [org.wildfly.extension.io] (ServerService Thread Pool -- 36) WFLYIO001: Worker 'default' has auto-configured to 4 core threads with 32 task threads based on your 2 available processors
 [0m [0m01:08:08,910 INFO  [org.jboss.as.clustering.infinispan] (ServerService Thread Pool -- 37) WFLYCLINF0001: Activating Infinispan subsystem.
 [0m [0m01:08:08,954 INFO  [org.jboss.as.naming] (ServerService Thread Pool -- 45) WFLYNAM0001: Activating Naming Subsystem
 [0m [0m01:08:08,956 INFO  [org.jboss.as.security] (ServerService Thread Pool -- 52) WFLYSEC0002: Activating Security Subsystem
 [0m [33m01:08:08,959 WARN  [org.jboss.as.txn] (ServerService Thread Pool -- 53) WFLYTX0013: Node identifier property is set to the default value. Please make sure it is unique.
 [0m [0m01:08:08,960 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 54) WFLYUT0003: Undertow 1.3.0.Final starting
 [0m [0m01:08:08,962 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 54) WFLYUT0014: Creating file handler for path '/opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/basic/target/jbossas/welcome-content' with options [directory-listing: 'false', follow-symlink: 'false', case-sensitive: 'true', safe-symlink-paths: '[]']
 [0m [0m01:08:08,975 INFO  [org.jboss.as.webservices] (ServerService Thread Pool -- 55) WFLYWS0002: Activating WebServices Extension
 [0m [0m01:08:08,988 INFO  [org.jboss.as.connector] (MSC service thread 1-1) WFLYJCA0009: Starting JCA Subsystem (WildFly/IronJacamar 1.3.1.Final)
 [0m [0m01:08:08,989 INFO  [org.jboss.as.security] (MSC service thread 1-4) WFLYSEC0001: Current PicketBox version=4.9.3.Final
 [0m [0m01:08:08,989 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-4) WFLYUT0003: Undertow 1.3.0.Final starting
 [0m [0m01:08:08,993 INFO  [org.jboss.as.connector.deployers.jdbc] (MSC service thread 1-2) WFLYJCA0018: Started Driver service with driver-name = h2
 [0m [0m01:08:08,993 INFO  [org.jboss.as.naming] (MSC service thread 1-2) WFLYNAM0003: Starting Naming Service
 [0m [0m01:08:08,993 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-3) WFLYSRV0027: Starting deployment of "arquillian-service" (runtime-name: "arquillian-service")
 [0m [0m01:08:08,994 INFO  [org.jboss.as.ejb3] (MSC service thread 1-1) WFLYEJB0482: Strict pool mdb-strict-max-pool is using a max instance size of 8 (per class), which is derived from the number of CPUs on this host.
 [0m [0m01:08:08,994 INFO  [org.jboss.as.ejb3] (MSC service thread 1-1) WFLYEJB0481: Strict pool slsb-strict-max-pool is using a max instance size of 32 (per class), which is derived from thread worker pool sizing.
 [0m [0m01:08:08,994 INFO  [org.jboss.as.mail.extension] (MSC service thread 1-1) WFLYMAIL0001: Bound mail session [java:jboss/mail/Default]
 [0m [0m01:08:08,999 INFO  [org.jboss.as.server.deployment.scanner] (MSC service thread 1-3) WFLYDS0013: Started FileSystemDeploymentService for directory /opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/basic/target/jbossas/standalone/deployments


```

--- a/undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentInfoService.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentInfoService.java
@@ -150,7 +150,8 @@ import org.wildfly.extension.undertow.security.SecurityContextAssociationHandler
 import org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction;
 import org.wildfly.extension.undertow.security.jacc.JACCAuthorizationManager;
 import org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler;
-import org.wildfly.extension.undertow.security.jaspi.JASPIAuthenticationMechanism;
+import org.wildfly.extension.undertow.security.jaspi.JASPICAuthenticationMechanism;
+import org.wildfly.extension.undertow.security.jaspi.JASPICSecureResponseHandler;
 import org.wildfly.extension.undertow.security.jaspi.JASPICSecurityContextFactory;
 import org.wildfly.extension.undertow.session.CodecSessionConfigWrapper;
 import org.wildfly.extension.undertow.session.SharedSessionManagerConfig;
@@ -452,7 +453,7 @@ public class UndertowDeploymentInfoService implements Service<DeploymentInfo> {
     }
 
     /**
-     * <p>Adds to the deployment the {@link JASPIAuthenticationMechanism}, if necessary. The handler will be added if the security domain
+     * <p>Adds to the deployment the {@link org.wildfly.extension.undertow.security.jaspi.JASPICAuthenticationMechanism}, if necessary. The handler will be added if the security domain
      * is configured with JASPI authentication.</p>
      *
      * @param deploymentInfo
@@ -469,8 +470,9 @@ public class UndertowDeploymentInfoService implements Service<DeploymentInfo> {
             if (loginConfig != null && loginConfig.getAuthMethods().size() > 0) {
                 authMethod = loginConfig.getAuthMethods().get(0).getName();
             }
-            deploymentInfo.setJaspiAuthenticationMechanism(new JASPIAuthenticationMechanism(securityDomain, authMethod));
+            deploymentInfo.setJaspiAuthenticationMechanism(new JASPICAuthenticationMechanism(securityDomain, authMethod));
             deploymentInfo.setSecurityContextFactory(new JASPICSecurityContextFactory(this.securityDomain));
+            deploymentInfo.addOuterHandlerChainWrapper(next -> new JASPICSecureResponseHandler(next));
         }
     }
 

--- a/undertow/src/main/java/org/wildfly/extension/undertow/logging/UndertowLogger.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/logging/UndertowLogger.java
@@ -331,4 +331,8 @@ public interface UndertowLogger extends BasicLogger {
     @LogMessage(level = ERROR)
     @Message(id = 78, value = "Failed to register management view for websocket %s at %s")
     void failedToRegisterWebsocket(Class endpoint, String path, @Cause Exception e);
+
+    @LogMessage(level = ERROR)
+    @Message(id = 77, value = "Error invoking secure response")
+    void errorInvokingSecureResponse(@Cause Exception e);
 }

--- a/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICAuthenticationMechanism.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICAuthenticationMechanism.java
@@ -25,12 +25,10 @@ import io.undertow.security.api.AuthenticatedSessionManager;
 import io.undertow.security.api.AuthenticationMechanism;
 import io.undertow.security.api.SecurityContext;
 import io.undertow.security.idm.Account;
-import io.undertow.server.ConduitWrapper;
 import io.undertow.server.HttpServerExchange;
 import io.undertow.servlet.handlers.ServletRequestContext;
 import io.undertow.util.AttachmentKey;
 
-import io.undertow.util.ConduitFactory;
 import io.undertow.util.StatusCodes;
 import org.jboss.security.SecurityConstants;
 import org.jboss.security.auth.callback.JBossCallbackHandler;
@@ -55,7 +53,6 @@ import org.jboss.security.auth.callback.JASPICallbackHandler;
 import org.jboss.security.identity.Role;
 import org.jboss.security.identity.RoleGroup;
 import org.wildfly.extension.undertow.security.UndertowSecurityAttachments;
-import org.xnio.conduits.StreamSinkConduit;
 
 /**
  * <p>
@@ -65,24 +62,24 @@ import org.xnio.conduits.StreamSinkConduit;
  * @author Pedro Igor
  * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
  */
-public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
+public class JASPICAuthenticationMechanism implements AuthenticationMechanism {
 
 
-    private static final String JASPI_HTTP_SERVLET_LAYER = "HttpServlet";
+    static final String JASPI_HTTP_SERVLET_LAYER = "HttpServlet";
     private static final String MECHANISM_NAME = "JASPIC";
     private static final String JASPI_AUTH_TYPE = "javax.servlet.http.authType";
     private static final String JASPI_REGISTER_SESSION = "javax.servlet.http.registerSession";
 
     public static final AttachmentKey<HttpServerExchange> HTTP_SERVER_EXCHANGE_ATTACHMENT_KEY = AttachmentKey.create(HttpServerExchange.class);
     public static final AttachmentKey<SecurityContext> SECURITY_CONTEXT_ATTACHMENT_KEY = AttachmentKey.create(SecurityContext.class);
-    private static final AttachmentKey<Boolean> ALREADY_WRAPPED = AttachmentKey.create(Boolean.class);
+
     public static final AttachmentKey<Boolean> AUTH_RUN = AttachmentKey.create(Boolean.class);
     public static final int DEFAULT_ERROR_CODE = StatusCodes.UNAUTHORIZED;
 
     private final String securityDomain;
     private final String configuredAuthMethod;
 
-    public JASPIAuthenticationMechanism(final String securityDomain, final String configuredAuthMethod) {
+    public JASPICAuthenticationMechanism(final String securityDomain, final String configuredAuthMethod) {
         this.securityDomain = securityDomain;
         this.configuredAuthMethod = configuredAuthMethod;
     }
@@ -95,7 +92,7 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
         final GenericMessageInfo messageInfo = createMessageInfo(exchange, sc);
         final String applicationIdentifier = buildApplicationIdentifier(requestContext);
         final JASPICallbackHandler cbh = new JASPICallbackHandler();
-
+        exchange.putAttachment(JASPICContext.ATTACHMENT_KEY, new JASPICContext(messageInfo, sam, cbh));
         UndertowLogger.ROOT_LOGGER.debugf("validateRequest for layer [%s] and applicationContextIdentifier [%s]", JASPI_HTTP_SERVLET_LAYER, applicationIdentifier);
 
         Account cachedAccount = null;
@@ -157,8 +154,6 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
         servletRequestContext.setServletRequest((HttpServletRequest) messageInfo.getRequestMessage());
         servletRequestContext.setServletResponse((HttpServletResponse) messageInfo.getResponseMessage());
 
-        secureResponse(exchange, sam, messageInfo, cbh);
-
         return outcome;
 
     }
@@ -168,15 +163,11 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
         return new ChallengeResult(true);
     }
 
-    private boolean wasAuthExceptionThrown(HttpServerExchange exchange) {
-        return exchange.getAttachment(UndertowSecurityAttachments.SECURITY_CONTEXT_ATTACHMENT).getData().get(AuthException.class.getName()) != null;
-    }
-
     private JASPIServerAuthenticationManager createJASPIAuthenticationManager() {
         return new JASPIServerAuthenticationManager(this.securityDomain, new JBossCallbackHandler());
     }
 
-    private String buildApplicationIdentifier(final ServletRequestContext attachment) {
+    static String buildApplicationIdentifier(final ServletRequestContext attachment) {
         ServletRequest servletRequest = attachment.getServletRequest();
         return servletRequest.getServletContext().getVirtualServerName() + " " + servletRequest.getServletContext().getContextPath();
     }
@@ -236,32 +227,6 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
         return new AccountImpl(userPrincipal, stringRoles, credential, original);
     }
 
-    private void secureResponse(final HttpServerExchange exchange, final JASPIServerAuthenticationManager sam, final GenericMessageInfo messageInfo, final JASPICallbackHandler cbh) {
-        if(exchange.getAttachment(ALREADY_WRAPPED) != null || exchange.isResponseStarted()) {
-            return;
-        }
-        exchange.putAttachment(ALREADY_WRAPPED, true);
-        // we add a response wrapper to properly invoke the secureResponse, after processing the destination
-        exchange.addResponseWrapper(new ConduitWrapper<StreamSinkConduit>() {
-            @Override
-            public StreamSinkConduit wrap(final ConduitFactory<StreamSinkConduit> factory, final HttpServerExchange exchange) {
-                ServletRequestContext requestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-                String applicationIdentifier = buildApplicationIdentifier(requestContext);
-
-                if (!wasAuthExceptionThrown(exchange)) {
-                    UndertowLogger.ROOT_LOGGER.debugf("secureResponse for layer [%s] and applicationContextIdentifier [%s].", JASPI_HTTP_SERVLET_LAYER, applicationIdentifier);
-                    sam.secureResponse(messageInfo, new Subject(), JASPI_HTTP_SERVLET_LAYER, applicationIdentifier, cbh);
-
-                    // A SAM can unwrap the HTTP request/response objects - update the servlet request context with the values found in the message info.
-                    ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-                    servletRequestContext.setServletRequest((HttpServletRequest) messageInfo.getRequestMessage());
-                    servletRequestContext.setServletResponse((HttpServletResponse) messageInfo.getResponseMessage());
-                }
-                return factory.create();
-            }
-        });
-    }
-
     /**
      * <p>The authentication is mandatory if the servlet has http constraints (eg.: {@link
      * javax.servlet.annotation.HttpConstraint}).</p>
@@ -275,5 +240,10 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
 
     private boolean statusIndicatesError(HttpServerExchange exchange) {
         return exchange.getResponseCode() != StatusCodes.OK;
+    }
+
+
+    static boolean wasAuthExceptionThrown(HttpServerExchange exchange) {
+        return exchange.getAttachment(UndertowSecurityAttachments.SECURITY_CONTEXT_ATTACHMENT).getData().get(AuthException.class.getName()) != null;
     }
 }

--- a/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICContext.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICContext.java
@@ -1,0 +1,37 @@
+package org.wildfly.extension.undertow.security.jaspi;
+
+import io.undertow.util.AttachmentKey;
+import org.jboss.security.auth.callback.JASPICallbackHandler;
+import org.jboss.security.plugins.auth.JASPIServerAuthenticationManager;
+
+import javax.security.auth.message.MessageInfo;
+
+/**
+ * @author Stuart Douglas
+ */
+public class JASPICContext {
+
+    public static final AttachmentKey<JASPICContext> ATTACHMENT_KEY = AttachmentKey.create(JASPICContext.class);
+
+    private final MessageInfo messageInfo;
+    private final JASPIServerAuthenticationManager sam;
+    private final JASPICallbackHandler cbh;
+
+    public JASPICContext(MessageInfo messageInfo, JASPIServerAuthenticationManager sam, JASPICallbackHandler cbh) {
+        this.messageInfo = messageInfo;
+        this.sam = sam;
+        this.cbh = cbh;
+    }
+
+    public MessageInfo getMessageInfo() {
+        return messageInfo;
+    }
+
+    public JASPIServerAuthenticationManager getSam() {
+        return sam;
+    }
+
+    public JASPICallbackHandler getCbh() {
+        return cbh;
+    }
+}

--- a/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICSecureResponseHandler.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICSecureResponseHandler.java
@@ -1,0 +1,48 @@
+package org.wildfly.extension.undertow.security.jaspi;
+
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.servlet.handlers.ServletRequestContext;
+import org.wildfly.extension.undertow.logging.UndertowLogger;
+
+import javax.security.auth.Subject;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * @author Stuart Douglas
+ */
+public class JASPICSecureResponseHandler implements HttpHandler {
+
+    private final HttpHandler next;
+
+    public JASPICSecureResponseHandler(HttpHandler next) {
+        this.next = next;
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        try {
+            next.handleRequest(exchange);
+        } finally {
+            try {
+                JASPICContext context = exchange.getAttachment(JASPICContext.ATTACHMENT_KEY);
+                ServletRequestContext requestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+                String applicationIdentifier = JASPICAuthenticationMechanism.buildApplicationIdentifier(requestContext);
+
+                if (!JASPICAuthenticationMechanism.wasAuthExceptionThrown(exchange)) {
+                    UndertowLogger.ROOT_LOGGER.debugf("secureResponse for layer [%s] and applicationContextIdentifier [%s].", JASPICAuthenticationMechanism.JASPI_HTTP_SERVLET_LAYER, applicationIdentifier);
+                    context.getSam().secureResponse(context.getMessageInfo(), new Subject(), JASPICAuthenticationMechanism.JASPI_HTTP_SERVLET_LAYER, applicationIdentifier, context.getCbh());
+
+                    // A SAM can unwrap the HTTP request/response objects - update the servlet request context with the values found in the message info.
+                    ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+                    servletRequestContext.setServletRequest((HttpServletRequest) context.getMessageInfo().getRequestMessage());
+                    servletRequestContext.setServletResponse((HttpServletResponse) context.getMessageInfo().getResponseMessage());
+                }
+            } catch (Exception e) {
+                UndertowLogger.ROOT_LOGGER.errorInvokingSecureResponse(e);
+            }
+        }
+    }
+
+}

--- a/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/modules/HTTPSchemeServerAuthModule.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/modules/HTTPSchemeServerAuthModule.java
@@ -27,7 +27,7 @@ import io.undertow.security.api.SecurityContext;
 import io.undertow.server.HttpServerExchange;
 import io.undertow.servlet.handlers.ServletRequestContext;
 import org.wildfly.extension.undertow.logging.UndertowLogger;
-import org.wildfly.extension.undertow.security.jaspi.JASPIAuthenticationMechanism;
+import org.wildfly.extension.undertow.security.jaspi.JASPICAuthenticationMechanism;
 
 import javax.security.auth.Subject;
 import javax.security.auth.callback.CallbackHandler;
@@ -77,8 +77,8 @@ public class HTTPSchemeServerAuthModule implements ServerAuthModule {
     @Override
     public AuthStatus validateRequest(MessageInfo messageInfo, Subject clientSubject, Subject serviceSubject)
             throws AuthException {
-        HttpServerExchange exchange = (HttpServerExchange) messageInfo.getMap().get(JASPIAuthenticationMechanism.HTTP_SERVER_EXCHANGE_ATTACHMENT_KEY);
-        SecurityContext securityContext = (SecurityContext) messageInfo.getMap().get(JASPIAuthenticationMechanism.SECURITY_CONTEXT_ATTACHMENT_KEY);
+        HttpServerExchange exchange = (HttpServerExchange) messageInfo.getMap().get(JASPICAuthenticationMechanism.HTTP_SERVER_EXCHANGE_ATTACHMENT_KEY);
+        SecurityContext securityContext = (SecurityContext) messageInfo.getMap().get(JASPICAuthenticationMechanism.SECURITY_CONTEXT_ATTACHMENT_KEY);
         ServletRequestContext src = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
         List<AuthenticationMechanism> mechanisms = src.getDeployment().getAuthenticationMechanisms();
 

